Scala 3: fails to wire constructor param with implicit lookup

Following discussion on: https://softwaremill.community/t/scala-3-fails-to-wire-constructor-param-with-implicit-lookup/346

The below example fails to compile with Scala 3. But works with Scala 2.x

Scala version: 3.3.1
macwire version: 2.5.9

```scala
import com.softwaremill.macwire.wire

type DBIO[T]
class ProductService[DB[_]]

trait MainModule {
  lazy val productService = wire[ProductService[DBIO]]
}
```

### Error log

```
Exception occurred while executing macro expansion.
java.lang.AssertionError: assertion failed
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
	at scala.quoted.runtime.impl.QuotesImpl$reflect$Ref$.apply(QuotesImpl.scala:444)
	at scala.quoted.runtime.impl.QuotesImpl$reflect$Ref$.apply(QuotesImpl.scala:443)
	at com.softwaremill.macwire.internals.ConstructorCrimper.paramType(ConstructorCrimper.scala:92)
	at com.softwaremill.macwire.internals.ConstructorCrimper.wireConstructorParamsWithImplicitLookups$$anonfun$4$$anonfun$3(ConstructorCrimper.scala:75)
	at scala.collection.immutable.List.map(List.scala:246)
	at com.softwaremill.macwire.internals.ConstructorCrimper.wireConstructorParamsWithImplicitLookups$$anonfun$1(ConstructorCrimper.scala:75)
	at scala.collection.immutable.List.map(List.scala:246)
	at com.softwaremill.macwire.internals.ConstructorCrimper.wireConstructorParamsWithImplicitLookups(ConstructorCrimper.scala:76)
	at com.softwaremill.macwire.internals.ConstructorCrimper.constructorArgs$$anonfun$3$$anonfun$1(ConstructorCrimper.scala:57)
	at scala.Option.map(Option.scala:242)
	at com.softwaremill.macwire.internals.ConstructorCrimper.constructorArgs$$anonfun$2(ConstructorCrimper.scala:57)
	at com.softwaremill.macwire.internals.Logger.withBlock(Logger.scala:15)
	at com.softwaremill.macwire.internals.ConstructorCrimper.constructorArgs(ConstructorCrimper.scala:58)
	at com.softwaremill.macwire.internals.ConstructorCrimper.constructorTree$$anonfun$5$$anonfun$3(ConstructorCrimper.scala:63)
	at scala.Option.flatMap(Option.scala:283)
	at com.softwaremill.macwire.internals.ConstructorCrimper.constructorTree$$anonfun$2(ConstructorCrimper.scala:67)
	at com.softwaremill.macwire.internals.Logger.withBlock(Logger.scala:15)
	at com.softwaremill.macwire.internals.ConstructorCrimper.constructorTree(ConstructorCrimper.scala:68)
	at com.softwaremill.macwire.MacwireMacros$.wire(MacwireMacros.scala:54)
	at com.softwaremill.macwire.MacwireMacros$.wireImpl(MacwireMacros.scala:14)
```

--- a/macros/src/main/scala-3/com/softwaremill/macwire/internals/ConstructorCrimper.scala
+++ b/macros/src/main/scala-3/com/softwaremill/macwire/internals/ConstructorCrimper.scala
@@ -51,7 +51,28 @@ private[macwire] class ConstructorCrimper[Q <: Quotes, T: Type](using val q: Q)(
     ctor
   }
 
-  lazy val constructorParamLists: Option[List[List[Symbol]]] = constructor.map(_.paramSymss)
+  private def constructorParamTypes(ctorType: TypeRepr): List[List[TypeRepr]] = {
+    ctorType match {
+      case MethodType(_, paramTypes, retType) =>
+        paramTypes.map(_.widen.simplified) :: constructorParamTypes(retType.simplified)
+      case _ =>
+        Nil
+    }
+  }
+
+  lazy val constructorParamLists: Option[List[List[(Symbol, TypeRepr)]]] = {
+    constructor.map { c =>
+      // paramSymss contains both type arg symbols (generic types) and value arg symbols
+      val symLists = c.paramSymss.filter(_.forall(!_.isTypeDef))
+      val ctorType =
+        if (targetType.typeArgs.isEmpty) targetType.memberType(c)
+        else targetType.memberType(c).appliedTo(targetType.typeArgs)
+      val typeLists = constructorParamTypes(ctorType)
+      symLists.zip(typeLists).map { case (syms, tpes) =>
+        syms.zip(tpes)
+      }
+    }
+  }
 
   lazy val constructorArgs: Option[List[List[Term]]] = log.withBlock("Looking for targetConstructor arguments") {
     constructorParamLists.map(wireConstructorParamsWithImplicitLookups)
@@ -62,35 +83,28 @@ private[macwire] class ConstructorCrimper[Q <: Quotes, T: Type](using val q: Q)(
       constructorValue <- constructor
       constructorArgsValue <- constructorArgs
     } yield {
-      val constructionMethodTree: Term = Select(New(TypeIdent(targetType.typeSymbol)), constructorValue)
+      val constructionMethodTree: Term = {
+        val ctor = Select(New(TypeIdent(targetType.typeSymbol)), constructorValue)
+        if (targetType.typeArgs.isEmpty) ctor else ctor.appliedToTypes(targetType.typeArgs)
+      }
       constructorArgsValue.foldLeft(constructionMethodTree)((acc: Term, args: List[Term]) => Apply(acc, args))
     }
   }
 
-  def wireConstructorParams(paramLists: List[List[Symbol]]): List[List[Term]] =
-    paramLists.map(_.map(p => dependencyResolver.resolve(p, /*SI-4751*/ paramType(p))))
+  def wireConstructorParams(paramLists: List[List[(Symbol, TypeRepr)]]): List[List[Term]] =
+    paramLists.map(_.map(p => dependencyResolver.resolve(p._1, /*SI-4751*/ p._2)))
 
-  def wireConstructorParamsWithImplicitLookups(paramLists: List[List[Symbol]]): List[List[Term]] = paramLists.map {
-    case params if params.forall(_.flags is Flags.Implicit) => params.map(resolveImplicitOrFail)
-    case params => params.map(p => dependencyResolver.resolve(p, /*SI-4751*/ paramType(p)))
-  }
+  def wireConstructorParamsWithImplicitLookups(paramLists: List[List[(Symbol, TypeRepr)]]): List[List[Term]] =
+    paramLists.map {
+      case params if params.forall(_._1.flags is Flags.Implicit) => params.map(resolveImplicitOrFail)
+      case params => params.map(p => dependencyResolver.resolve(p._1, /*SI-4751*/ p._2))
+    }
 
-  private def resolveImplicitOrFail(param: Symbol): Term = Implicits.search(paramType(param)) match {
-    case iss: ImplicitSearchSuccess => iss.tree
-    case isf: ImplicitSearchFailure => report.throwError(s"Failed to resolve an implicit for [$param].")
-  }
-
-  private def paramType(param: Symbol): TypeRepr = {
-    // val (sym: Symbol, tpeArgs: List[Type]) = targetTypeD match {
-    //   case TypeRef(_, sym, tpeArgs) => (sym, tpeArgs)
-    //   case t => abort(s"Target type not supported for wiring: $t. Please file a bug report with your use-case.")
-    // }
-    // val pTpe = param.signature.substituteTypes(sym.asClass.typeParams, tpeArgs)
-    // if (param.asTerm.isByNameParam) pTpe.typeArgs.head else pTpe
-
-    //FIXME assertion error in test inheritanceHKT.success, selfTypeHKT.success
-    Ref(param).tpe.widen
-  }
+  private def resolveImplicitOrFail(param: Symbol, paramType: TypeRepr): Term =
+    Implicits.search(paramType) match {
+      case iss: ImplicitSearchSuccess => iss.tree
+      case isf: ImplicitSearchFailure => report.throwError(s"Failed to resolve an implicit for [$param].")
+    }
 
   /** In some cases there is one extra (phantom) constructor. This happens when extended trait has implicit param:
     *

--- a/tests/src/test/resources/test-cases/simpleGenericClass.success
+++ b/tests/src/test/resources/test-cases/simpleGenericClass.success
@@ -1,0 +1,14 @@
+#include commonHKTClasses
+
+case class LocalA[X]()
+case class LocalB[X](la: LocalA[X])
+
+object Test {
+  val la = wire[LocalA[Int]]
+  val lb = wire[LocalB[Int]]
+  val a = wire[A[IO]]
+  val b = wire[B[IO]]
+}
+
+require(Test.lb.la == Test.la)
+require(Test.b.a == Test.a)

